File: shoring/compat/src/main/java/com/panopset/compat/PersistentMapFile.java
package com.panopset.compat;

import java.io.File;
import java.io.IOException;
import java.util.Map;
import java.util.Properties;
import java.util.Set;

public class PersistentMapFile implements PersistentMap {

	public PersistentMapFile() {
	}

	public PersistentMapFile(File newFile) {
		setFile(newFile);
	}

	public void setFile(final File newFile) {
		file = newFile;
	}

	public void put(String key, String value) {
		if (key == null) {
			Logop.warn("key is null");
			return;
		}
		if (value == null) {
			Logop.warn("value is null");
			return;
		}
		try {
			getProps().setProperty(key, value);
		} catch (IOException ex) {
			Logop.error(ex);
		}
	}

	public Set<Map.Entry<Object, Object>> entrySet() throws IOException {
		return getProps().entrySet();
	}

	@Override
	public String get(String key, String defaultValue) throws IOException {
		if (!Stringop.isPopulated(key)) {
			return defaultValue;
		}
		String rtn = getProps().getProperty(key);
		if (rtn == null) {
			return defaultValue;
		}
		return rtn;
	}

	public String get(String key) throws IOException {
		return get(key, "");
	}

	public void purge() {
		if (Fileop.fileExists(file)) {
			file.delete();
		}
	}

	public void flush() {
		try {
			Propop.save(getProps(), file);
		} catch (IOException ex) {
			Logop.error(ex);
		}
	}

	private Properties getProps() throws IOException {
		if (props == null) {
			props = Propop.load(file);
		}
		return props;
	}

	private File file;
	private Properties props;

	public void load() {
		props = null;
	}

	public String getFileName() {
		if (file == null) {
			return "";
		}
		return Fileop.getCanonicalPath(file);
	}
}

File: code/shoring/compat/src/main/kotlin/com/panopset/compat/PersistentMapFile.kt
package com.panopset.compat

import java.io.File
import java.io.IOException
import java.util.*

class PersistentMapFile(newFile: File) : PersistentMap {

    fun setNewFile(newFile: File) {
        file = newFile
    }

    override fun put(key: String, value: String) {
        try {
            getProps().setProperty(key, value)
        } catch (ex: IOException) {
            Logz.errorEx(ex)
        }
    }

    fun entrySet(): Set<Map.Entry<Any, Any>> {
        return getProps().entries
    }

    override fun getMapValue(key: String, dft: String): String {
        return if (Stringop.isPopulated(key)) {
            getProps().getProperty(key) ?: return dft
        } else {
            dft
        }
    }

    override fun getMapValue(key: String): String {
        return getMapValue( key, "")
    }

    override fun containsKey(key: String): Boolean {
        return getProps().containsKey(key)
    }

    fun purge() {
        if (Fileop.fileExists(file)) {
            file.delete()
        }
    }

    fun flush() {
        try {
            Propop.save(getProps(), file)
        } catch (ex: IOException) {
            Logz.errorEx(ex)
        }
    }

    private fun getProps(): Properties {
        if (props.isEmpty) {
            load()
        }
        return props
    }

    var file: File = File("")

    private var props: Properties = Properties()
    fun load() {
        props.clear()
        props = Propop.load(file)
    }

    init {
        setNewFile(newFile)
    }
}
